### ISSUE TITLE
False == 0 fails

### ISSUE THREAD

**Issue by JohnDenker** [kovidgoyal/rapydscript-ng#38]
**Recipe:**
Consider the following seemingly-innocent program:
```
a = 3==4
print("yyyyy", a, 0, a==0)
```
**Desired and expected output:**
``yyyyy False 0 True``

This works as expected under real python from the command line.
The corresponding javascript also works as expected. However...

**Observed rapydscript behavior:**
``yyyyy false 0 false``

This is 100% reproducible chez moi.  I am using the latest rapydscript-ng, from a fresh git-pull.  FWIW the old non-ng version exhibits the same misbehavior.

This is more important than it might seem. The example above is highly simplified, to facilitate debugging. The same issue affects more-complicated code in ways that are much harder to understand. This wastes significant amounts of programmer time.

This started out as BruceSherwood/glowscript#60 but is hereby pushed upstream.

**Comment by kovidgoyal** [comment]
RapydScript does not do operator overloading (with a minor caveat). The performance implications of operator overloading are too severe. As such the result of == is the same as the result of === in JS and in JS 0 !== false

See the gotchas section of the readme.

**Comment by kovidgoyal** [comment]
And let me just note that I am totally unconvinced that `0 == False` is important. Having worked on over a million lines of python in my career, I have never come across the need for that. 

`bool(0) == False` on the other hand is important. so that

`if a:`

has the same result for both `a=0` and `a=False`. And this is the case in rapydscript.

**Comment by JohnDenker** [comment]
1.  As to importance and prevalence, I just did some checking.  I found that the subexpression ``== 0`` occurred 37,000 times in 21 million lines of code.  It occurred in 9000 out of 35000 files.  That's more than 25% of the files.  So it would seem that using ``== 0`` is not particularly unreasonable.

2.  Also note that as mentioned in the original report, the given example is highly simplified.  Usually it is considered helpful and polite to submit simple examples, rather than submitting a huge pile of real-world code and asking the developers to wade through it.  We agree that users do not normally write ``a = 3==4`` and ``a==0`` on successive lines, but that's not the relevant use-case.  This bug originated in real-world code, buried in a sub-sub-sub-sub-function, where the ``==`` operator was applied to the copy of a copy of a copy of a copy of a variable.  At runtime, sometimes the passed-in variable was a boolean and sometimes it wasn't.  Often it was a member of an opaque class, so that the user had no way of knowing (short of experimentation) whether any given instance was a boolean or not.  In any case, the point remains:  *it should not have mattered*.  Asking users to be defensive about ``== 0`` would be unreasonable.

3. I have read the gotcha list at https://www.npmjs.com/package/rapydscript-ng#gotchas
I do not see anything there that helps the user understand this bug, or explains how to avoid this bug. 

- The distinction between python and javascript is mentioned there, but I don't see how that is relevant, because ``== 0`` works as expected in javascript as well as python.
-  ``True`` and ``False`` are mentioned there, but only in connection with the length of lists.  Neither ``==`` nor ``===`` is mentioned there.
- I don't see how ``===`` is relevant, because the user wrote ``==`` not ``===``.
- I don't see how operator overloading is relevant.  Keep in mind that the user wrote ``== 0`` which works as expected in a very wide range of languages, including javascript as well as python.  No overloading required.

**Comment by kovidgoyal** [comment]
> 1.  As to importance and prevalence, I just did some checking.  I found that the subexpression ``== 0`` occurred 37,000 times in 21 million lines of code.  It occurred in 9000 out of 35000 files.  That's more than 25% of the files.  So it would seem that using ``== 0`` is not particularly unreasonable.

I did not say == 0 is uncommon, I said False == 0 is uncommon.

> 
> 2.  Also note that as mentioned in the original report, the given example is highly simplified.  Usually it is considered helpful and polite to submit simple examples, rather than submitting a huge pile of real-world code and asking the developers to wade through it.  We agree that users do not normally write ``a = 3==4`` and ``a==0`` on successive lines, but that's not the relevant use-case.  This bug originated in real-world code, buried in a sub-sub-sub-sub-function, where the ``==`` operator was applied to the copy of a copy of a copy of a copy of a variable.  At runtime, sometimes the passed-in variable was a boolean and sometimes it wasn't.  Often it was a member of an opaque class, so that the user had no way of knowing (short of experimentation) whether any given instance was a boolean or not.  In any case, the point remains:  *it should not have mattered*.  Asking users to be defensive about ``== 0`` would be unreasonable.

No it would not be unreasonable. Checking that something that could be
an integer or something "else" is == 0 is meaningless. The only
meaningful check you can make there is for the "truthiness"
of the variable. And as I said before that check works exactly as expected in 
RapydScript. If there exists code uses == 0 as a way to check for
truthiness, then the code is simply wrong and needs to be fixed.

> 
> 3. I have read the gotcha list at https://www.npmjs.com/package/rapydscript-ng#gotchas
> I do not see anything there that helps the user understand this bug, or explains how to avoid this bug. 

Read it again, this time a little more carefully.

**Comment by JohnDenker** [comment]
Using ``== 0`` to check for truthiness is not wrong in javascript.
Using ``== 0`` to check for truthiness is not wrong in python.
Using ``== 0`` to check for truthiness is not wrong in a whole lot of other languages besides.
The assertion that it is «simply wrong» is at best a novel idea.
The assertion that it is «simply wrong» is at best a matter of opinion.
Treating it as «wrong» serves no useful purpose; see below for details.

A my friend TIM TOWODI likes to say:  There Is More Than One Way Of Doing It.
In particular, in python ``X == 0`` works the same as ``not X`` no matter whether X is bool, int, or float.
Similarly in javascript, ``X == 0`` works the same as ``!X`` no matter whether X is boolean or number.
The same goes for a great many other languages.
I cannot imagine any good reason for changing it.  The cost of producing the expected behavior is small compared to the cost of explaining the unexpected behavior.

I have read the gotcha list at https://www.npmjs.com/package/rapydscript-ng#gotchas as carefully as humanly possible.  It still does not tell the user how to understand this bug or how to avoid this bug.

The assertion that rapydscript does not do overloading is inconsistent with observed reality.  I observe that python ``==`` already translates to javascript ``===`` overloaded onto ``ρσ_equals``.  So the question is not whether type-checking will be done;  the question is whether it will be done correctly.

In particular, replacing
``(a === b || typeof a === "object" && ρσ_equals(a, b))``
with something like
``(typeof a === "object" ? ρσ_equals(a, b) : a == b)``
should be more correct.  It should be equally fast (or slightly faster) in the case where the whole thing evaluates to false, and only slightly slower in the worst case.

Note that a modern compiler will often optimize out type-checks of this ilk, resulting in no performance penalty at all.

Last but not least, broad assertions about «unacceptable» performance penalties are almost always wrong.  As Knuth famously said:  
>  premature optimization is the root of all evil.

In part that's because in the tiny fraction of cases where any of this matters, the user can find the tiny bit of code that is performance-critical and hand-optimize the inner loop.  This includes rewriting it  in bare javascript if necessary.  This may involve removing the type-checking that rapydscript puts in.

To say the same thing the other way:  Correctness must come before speed.  There is no advantage to having a program that produces the wrong answer quickly.

**Comment by kovidgoyal** [comment]
> Using ``== 0`` to check for truthiness is not wrong in javascript.
> Using ``== 0`` to check for truthiness is not wrong in python.
> Using ``== 0`` to check for truthiness is not wrong in a whole lot of other languages besides.

It is wrong, just because it gives the right result in some cases does not make it
correct.

> The assertion that it is «simply wrong» is at best a novel idea.
> The assertion that it is «simply wrong» is at best a matter of opinion.

The assertion that is not wrong is at best an insane idea and a matter
of pretty idiotic opinion. One that stems from the legacy of languages
like C which had no boolean type and therefore used 0 and 1 as
stand-ins.

> Treating it as «wrong» serves no useful purpose; see below for details.

Yes it does. It improves performance. You want to claim it does not
provide some evidence, dont quote Knuth's famous cliche.

> 
> A my friend TIM TOWODI likes to say:  There Is More Than One Way Of Doing It.
> In particular, in python ``X == 0`` works the same as ``not X`` no matter whether X is bool, int, or float.
> Similarly in javascript, ``X == 0`` works the same as ``!X`` no matter whether X is boolean or number.

Sigh, go read up on truthiness.

In python:


```
if '' is False
'' != False
if [] is False
[] != False

```

There is absolutely no good reason why 0 should be treated differently
from '' and []

That it is treated differently is likely a historical accident for
compatibility with C like languages. Not a goal for which one should
sacrifice performance, code size and simplicity.

You disagree, feel free to fork RapydScript. You are not going to
convince me to degrade all RapydScript output to humor this asinine use case.

> The assertion that rapydscript does not do overloading is inconsistent with observed reality.  I observe that python ``==`` already translates to javascript ``===`` overloaded onto ``ρσ_equals``.  So the question is not whether type-checking will be done;  the question is whether it will be done correctly.

You conveniently ignored the bit where I said, with one small caveat
didn't you?

**Comment by kovidgoyal** [comment]
And let me add a few examples from python to illuminate why using == 0 to check for truthiness is wrong.

```py
0 != ''
bool(0) == bool('')
0 != []
bool(0) == bool([])
```
and more generally, for any type that overloads the == operator by declaring a __bool__ method, using `== 0` to test for truthiness will fail.

**Comment by JohnDenker** [comment]
Not all languages play by the same rules.  For example, in common lisp (+ 3 t) is an error.  However, the fact remains that javascript, and python, and a great many other languages do provide automatic conversion between bools and ints.

> The assertion that is not wrong is at best an insane idea and a matter of pretty idiotic opinion. 

Let's take a look at the ECMAScript® Language Specification (7ᵗʰ Edition / June 2016)

http://www.ecma-international.org/ecma-262/7.0/index.html#sec-type-conversion
**The ECMAScript language implicitly performs automatic type conversion as needed.**

http://www.ecma-international.org/ecma-262/7.0/index.html#sec-toboolean
**ToBoolean ( argument )**
**Number :: Return false if argument is +0, -0, or NaN; otherwise return true.**

http://www.ecma-international.org/ecma-262/7.0/index.html#sec-tonumber
**ToNumber ( argument )**
**Boolean :: Return 1 if argument is true. Return +0 if argument if false.**

This applies to everything, not limited to truthiness.  For example, ``Math.pow(3, 4==4)`` evaluates to 3, as it must, in accordance with the specification.

This is the «opinion» of a 55-year-old international standards body.  This is a «legacy» from June of 2016 ... although virtually identical language was present in the first edition, back in 1997:
http://www.ecma-international.org/publications/files/ECMA-ST-ARCH/ECMA-262,%201st%20edition,%20June%201997.pdf

If you think it is «insane» and «idiotic», please argue with Ecma, not with me.

**Comment by JohnDenker** [comment]
I might add that in common lisp, (+3 t) throws an error.

I suggest that if you really think ``a = 3==4;  a==0`` is somehow «wrong», it should throw an error.

In contrast, silently returning a numerically-incorrect answer is unhelpful.  It is not an appropriate way of punishing the unsuspecting user who wrote standards-compliant yet somehow allegedly «wrong» code.

**Comment by kovidgoyal** [comment]
Why on earth do you think the ECMA script standard has anything to do with RapydScript? I explained, in painful detail, to you why using == 0 as a test of truthiness is wrong in *python* and you come back with quoting the ecmascript standard -- I am done wasting time on this.

Good bye and good luck.

**Comment by JohnDenker** [comment]
> You are not going to convince me to degrade all RapydScript output to humor this asinine use case

What is the evidence that fixing this bug will degrade anything?  Please let's discuss facts not conjectures.

In particular, consider the attached program
[eq-timing.py.txt](https://github.com/kovidgoyal/rapydscript-ng/files/360545/eq-timing.py.txt)
It works fine under python2.7 and python3.  If you chop off the first few lines, it works fine under rapydscript.  It compares ints to ints, so everybody agrees what the answer should be.  The only question is the timing.

I observe 208.399 ns per once through the inner loop, using nodejs applied to the javascript emitted by rapydscript-ng.    The crucial javascript statement can be outlined as ``a===b || type === object && ρσ_equals(a,b)``

Now consider replacing that by ``type === object ? ρσ_equals(a,b) : a==b`` as I previously suggested.  I see no evidence that this is slower.  Indeed when I do this, the benchmark program reports 198 ns, which is very slightly faster.

To summarize: Treating ``bool == int`` as «wrong» serves no useful purpose.
Fixing this bug incurs no penalty, at runtime or otherwise, according to all evidence available to me.
Implementing the expected behavior is easier the explaining the unexpected non-standards-compliant behavior.

The bottom line is that the suggested scheme **gets the right answer** and is at least comparably efficient.

**Comment by JohnDenker** [comment]
Here is the corresponding quote from the python documentation:
    https://docs.python.org/2/library/stdtypes.html#boolean-values
**In numeric contexts (for example when used as the argument to an arithmetic operator), they behave like the integers 0 and 1, respectively.**

Programmers rely on this.  If you think it is «wrong» «idiotic» «insane» and «asinine» I suggest you report it to the folks at python.org.  Let me know how that goes.

FWIW I quoted the Ecma standard because I thought you were suggesting that the unexpected behavior was somehow attributable to the difference between python and javascript.  In any case, no matter which documentation you look at, the answer is the same.

**Comment by kovidgoyal** [comment]
Sigh. Once again, using == 0 is wrong for all the reasons I listed above. None of which you address. I long ago said that python does indeed allow 0 == False, presumably for historical compatibility with C like languages, that does not make it right. Go read what I wrote above, again. Using == 0 to test for truthiness is simply wrong. Admit it and move on.

Try addressing the actual arguments I am making instead of whatever argument you have unrolling in your own head. 

And by the way, changing the equality operator as you suggest would cause `1 == '1'` to return true. There is a reason RS uses the JS `===` operator and not the `==` operator. Go read up about the problems with the == operator in JS.

**Comment by kovidgoyal** [comment]
Oh and incidentally, running micro-benchmarks with an active JIT is worse than useless.

**Comment by JohnDenker** [comment]
I address the repeated argument about truthiness as follows: This bug is not about truthiness.  The ``==`` operator does not check for truthiness.  The user did not write or intend a check for truthiness.  If you want to discuss truthiness, let's discuss that in some other thread.

This bug has more to do with type conversions.  In particular, the free conversion between integers and booleans is a documented feature of both languages, both python and javascript.   One may opine about the origins of this feature, but the fact remains that type conversion is a feature of both languages.  Programmers rely on it.

In a comparison between False and zero, you can convert the LHS to an integer, or you can convert the RHS to boolean, and you get the same correct answer either way.  Programmers rely on this.

As for the observation that '"1" is considered equal to 1 in javascript, that is exactly the sort of thing that could reasonably be added to the gotcha discussion, as a reminder that RS is just a thin veneer on top of javascript.  It is understandable that some features will reflect javascript semantics, even though proper python semantics would be preferable.

In contrast, this bug concerns a situation where the code is standards-compliant in both languages and has the same semantics in both languages.  This is not explained on the gotcha list at https://www.npmjs.com/package/rapydscript-ng#gotchas and would be very hard to explain.  If a tradeoff must be made, a documented issue concerning strings and numbers would be vastly preferable to an undocumented issue concerning booleans and numbers.

To repeat:  Semantics that adheres to one standard or the other is understandable.  Semantics that violates both standards, on a topic where they agree, is much harder to understand.  

As for the benchmark numbers:  Imperfect data is better than no data.  If you have better data, please share it.   It seems *a priori* implausible that ``type === object ? ρσ_equals(a,b) : a==b`` would be appreciably less efficient that what RS is doing now.

I would also remark that the nodejs compiler is not stupid.  For the benchmark in question, RS+nodejs runs faster than real python2.7, python3, or ipython.  In fact, because nodejs does such aggressive optimization, writing meaningful benchmarks requires some expertise.  All of the per-iteration numbers I quoted are insensitive to modest changes in the number of iterations, indicating that any JIT overhead has been well amortized.  If you know a way of obtaining better data, please explain.

**Comment by kovidgoyal** [comment]
Sigh, I give up, feel free to keep on talking to yourself.

**Comment by JohnDenker** [comment]
> Sigh, go read up on truthiness.

I agree that reading the language specifications is a good idea in situations like this.

Just recently I was reading https://docs.python.org/3.1/library/stdtypes.html#numeric-types-int-float-complex where the second sentence says:   **Booleans are a subtype of integers.**

Python considers False == 0 to be a comparison between integers.  Booleans inherit comparison from integers.  They inherit a lot of other things, too;  for example:  ``(1==1) + (2==2)`` evaluates to 2.

Calling it «wrong in python» does not make it so.  Calling it «insane» and «idiotic» and «asinine» does not make it so.

If you wish to argue that booleans ought not be a subtype of integers, please argue with the folks at python.org, not with me.

Javascript arrives at the same destination via a different path, via automatic type conversions, as previously discussed.  The automatic conversion of integers to booleans, and vice versa, is mandatory.  It is not open to question.

**Comment by kovidgoyal** [comment]
I'm going to spell it out one last time for anyone else reading this bug report.

1) There is only one possible reason to do `a == 0` when a is of unknown type, and that is to check its truthiness. You trying to move the goalposts notwithstanding.

2) Using `a == 0` to check truthiness is **wrong**, because it fails when a is any of; None, empty list, empty tuple, empty set, empty dict, empty string, any user defined type with a `__bool__` method. 

3) Your proposed solution, to use the javascript `==` operator instead of the `===` operator is so stupid, it boggles the mind. You want to bring in all the infelicities of the JS == operator in order to enable `a == 0`, which is a stupid thing to do in the first place. And here is a link that explicates some of the problems with the `==` operator in JS. http://stackoverflow.com/questions/359494/does-it-matter-which-equals-operator-vs-i-use-in-javascript-comparisons

@JohnDenker Dont bother replying (although I suspect you cannot resist) I have blocked you.
